Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp

```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
     return maskedImage ? maskedImage : image;
 }
 
+String ImageDecoderCG::decodeUTI(const SharedBuffer& data) const
+{
+    auto uti = String(CGImageSourceGetType(m_nativeDecoder.get()));
+    if (uti != "public.heif"_s && uti != "public.heic"_s)
+        return uti;
+
+    if (data.size() < sizeof(unsigned)) {
+        ASSERT_NOT_REACHED();
+        return uti;
+    }
+
+    static constexpr auto ftypSignature = FourCC("ftyp");
+    static constexpr auto avifBrand = FourCC("avif");
+    static constexpr auto avisBrand = FourCC("avis");
+    
+    auto boxUnsigned = [&data](unsigned index) -> unsigned {
+        static constexpr bool littleEndian = false;
+        const unsigned* boxBytes = reinterpret_cast<const unsigned*>(data.data());
+        // Let's make all the numbers in big endian because FourCC are big endian.
+        return flipBytesIfLittleEndian(boxBytes[index], littleEndian);
+    };
+
+    auto checkAVIFBrand = [](unsigned brand) -> std::optional<String> {
+        if (brand == avifBrand)
+            return "public.avif"_s;
+        if (brand == avisBrand)
+            return "public.avis"_s;
+        return std::nullopt;
+    };
+
+    // HEIF/AVIF files start with an ftyp box
+    //
+    //  aligned(8) class GeneralTypeBox(code) extends Box(code) {
+    //      unsigned int(32)    major_brand;
+    //      unsigned int(32)    minor_version;
+    //      unsigned int(32)    compatible_brands[];    // to end of the box
+    //  }
+    //
+    // A box starts with 4 bytes for the size (including the length bytes).
+    // An image is considered an AVIF if "avif" or "avis" is present in its
+    // major_brand or in one of the compatible_brands.
+
+    // Get the box size. This size includes the length bytes.
+    unsigned boxUnsignedIndex = 0;
+    unsigned boxByteSize = boxUnsigned(boxUnsignedIndex++);
+    if (boxByteSize % 4 || boxByteSize > data.size())
+        return uti;
+
+    unsigned boxUnsignedSize = boxByteSize / sizeof(unsigned);
```

Comment:
Isn’t it meaningless to do math on a big-endian number?

Reply:
I am sorry for the confusion but I was confused myself by the little/big endian. Also the names of existing APIs do not help explain what I want to do. `flipBytesIfLittleEndian(num, false)` converts a big endian unsigned to match the Endianness of the current CPU architecture. So the math is correct in this case. And I fixed the comment that was saying "let's convert everything to big endian".